### PR TITLE
Create authorization service

### PR DESCRIPTION
**Pull request by KirillMeleshko004**

### CODE REVIEW

File: Shop.BLL/Exceptions/InternalExceptions/InternalException.cs [hunk 1]
```diff
@@ -0,0 +1,5 @@
+namespace Shop.BLL.Exceptions.InternalExceptions
+{
+    public class InternalException() :
+        Exception("Internal error occured during request processing");
```
Comment (Elizaveta564) [inline Shop.BLL/Exceptions/InternalExceptions/InternalException.cs]: you need to pass a variable, and write the message itself in a series like string message

File: Shop.BLL/Services/AuthService.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        {
+            _mapper = mapper;
+            _userManager = userManager;
+            _logger = logger;
+        }
+        public async Task<bool> AuthorizeUserAsync(UserRequestAuthorizationDto userRequestAuthorizationDto)
+        {
+            var user = await _userManager
+                .FindByEmailAsync(userRequestAuthorizationDto.Email);
+
```
Comment (Elizaveta564) [inline Shop.BLL/Services/AuthService.cs:31]: it is necessary to check whether such a user exists and if not, then throw the appropriate exception like if(user is null){...}

File: Shop.BLL/Services/AuthService.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            var res = user != null && await _userManager.CheckPasswordAsync(user,
+                userRequestAuthorizationDto.Password);
+
```
Comment (Elizaveta564) [inline Shop.BLL/Services/AuthService.cs:34]: the same, if(!checkPassword){....UnauthorizedAccessException....}

File: Shop.BLL/Services/AuthService.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            if (!res)
+            {
+                return false;
+            }
+
+            return true;
```
Comment (Elizaveta564) [inline Shop.BLL/Services/AuthService.cs]: in this method, you need to create a token and return it

File: Shop.BLL/Services/AuthService.cs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+                var errors = new StringBuilder();
+                foreach (var error in res.Errors)
+                {
+                    errors.Append($"{error.Code}:{error.Description}\n");
+                }
+
+                _logger.LogError("Error occured while creating user: {error}",
+                    errors);
+
+                throw new InternalException();
```
Comment (Elizaveta564) [inline Shop.BLL/Services/AuthService.cs]: it is better to use BadRequestException

**Review by Elizaveta564 — CHANGES_REQUESTED** [review]

File: Shop.BLL/Services/AuthService.cs (hunk 4 above)
Comment (KirillMeleshko004) [reply Shop.BLL/Services/AuthService.cs]: It was discussed at daily. Token issuing perfomed by OpenIdDict using SignIn method with OpenIddictServerAspNetCoreDefaults.AuthenticationScheme

File: Shop.BLL/Services/AuthService.cs (hunk 2 above)
Comment (KirillMeleshko004) [reply Shop.BLL/Services/AuthService.cs:31]: If notfound exception is thrown, global exception handler will capture exception and return 404, which let client know, that user with that name does not exist which is vulnerability. I suppose its better just return false to controller and then return Unauthorized

**Review by Elizaveta564 — APPROVED** [review]

**Pull request merged**